agasc/agasc.py

```diff
@@ -826,3 +833,128 @@ def set_star(star, name, value):
+    """
+    Write a HEALPix index table to an HDF5 file.
+
+    Parameters
+    ----------
+    filename : str
+        The path to the HDF5 file to write to.
+    healpix_index : astropy.table.Table
+        The HEALPix index table to write.
+    nside : int
+        The NSIDE parameter used to generate the HEALPix index.
+
+    Returns
+    -------
+    None
+    """
+    healpix_index_np = healpix_index.as_array()
+
+    with tables.open_file(filename, mode="a") as h5:
+        h5.create_table("/", "healpix_index", healpix_index_np, title="HEALPix index")
+        h5.root.healpix_index.attrs["nside"] = nside
+
+
+DTYPE = np.dtype(
+    [
+        ("AGASC_ID", np.int32),
+        ("RA", np.float64),
+        ("DEC", np.float64),
+        ("POS_ERR", np.int16),
+        ("POS_CATID", np.uint8),
+        ("EPOCH", np.float32),
+        ("PM_RA", np.int16),
+        ("PM_DEC", np.int16),
+        ("PM_CATID", np.uint8),
+        ("PLX", np.int16),
+        ("PLX_ERR", np.int16),
+        ("PLX_CATID", np.uint8),
+        ("MAG_ACA", np.float32),
+        ("MAG_ACA_ERR", np.int16),
+        ("CLASS", np.int16),
+        ("MAG", np.float32),
+        ("MAG_ERR", np.int16),
+        ("MAG_BAND", np.int16),
+        ("MAG_CATID", np.uint8),
+        ("COLOR1", np.float32),
+        ("COLOR1_ERR", np.int16),
+        ("C1_CATID", np.uint8),
+        ("COLOR2", np.float32),
+        ("COLOR2_ERR", np.int16),
+        ("C2_CATID", np.uint8),
+        ("RSV1", np.float32),
+        ("RSV2", np.int16),
+        ("RSV3", np.uint8),
+        ("VAR", np.int16),
+        ("VAR_CATID", np.uint8),
+        ("ASPQ1", np.int16),
+        ("ASPQ2", np.int16),
+        ("ASPQ3", np.int16),
+        ("ACQQ1", np.int16),
+        ("ACQQ2", np.int16),
+        ("ACQQ3", np.int16),
+        ("ACQQ4", np.int16),
+        ("ACQQ5", np.int16),
+        ("ACQQ6", np.int16),
+        ("XREF_ID1", np.int32),
+        ("XREF_ID2", np.int32),
+        ("XREF_ID3", np.int32),
+        ("XREF_ID4", np.int32),
+        ("XREF_ID5", np.int32),
+        ("RSV4", np.int16),
+        ("RSV5", np.int16),
+        ("RSV6", np.int16),
+    ]
+)
+
+
+class Order(Enum):
+    NONE = 1
+    DEC = 2
+    HEALPIX = 3
+
+
+def write_agasc(
+    filename: str, stars: np.ndarray, version: str, nside=64, order=Order.HEALPIX
+):
+    if stars.dtype != DTYPE:
+        cols = DTYPE.names
+        missing_keys = set(cols) - set(stars.dtype.names)
+        assert not missing_keys, f"Missing keys in stars: {missing_keys}"
+        stars = Table(stars)[cols].as_array().astype(DTYPE)
+    assert stars.dtype == DTYPE
+
+    if order == Order.DEC:
+        logger.info("Sorting on DEC column")
+        idx_sort = np.argsort(stars["DEC"])
+    elif order == Order.HEALPIX:
+        logger.info(
+            f"Creating healpix_index table for nside={nside} "
+            "and sorting by healpix index"
+        )
+        healpix_index, idx_sort = get_healpix_index_table(stars, nside)
```

Comment:
This is an opportunity to use the new `switch` statement in Python 3.10!

Reply:
As far as I understand it, the switch statement gains us nothing in terms of performance, and it actually makes the code slightly more verbose, so I'm not sure we want to use it.

Anyway, I changed it.